pkg/auth/auth.go

```diff
@@ -72,20 +91,199 @@ func CreateUser(c *gin.Context) {
+//	@Failure		409		{object}	models.LicenseError	"User already exists"
+//	@Router			/users/oidc [post]
+func CreateOidcUser(c *gin.Context) {
+	var tokenBody models.OidcUserCreate
+	if err := c.ShouldBindJSON(&tokenBody); err != nil {
+		er := models.LicenseError{
+			Status:    http.StatusBadRequest,
+			Message:   "invalid json body",
+			Error:     err.Error(),
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusBadRequest, er)
+		return
+	}
+
+	parsedToken, err := jwt.Parse(tokenBody.Token, Jwks.Keyfunc)
+	if err != nil || !parsedToken.Valid {
+		er := models.LicenseError{
+			Status:    http.StatusInternalServerError,
+			Message:   "Something went wrong",
+			Error:     err.Error(),
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusInternalServerError, er)
+		return
+	}
+
+	claims, ok := parsedToken.Claims.(jwt.MapClaims)
+	if !ok {
+		er := models.LicenseError{
+			Status:    http.StatusInternalServerError,
+			Message:   "Something went wrong",
+			Error:     "Please check your credentials and try again",
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusInternalServerError, er)
+		return
+	}
+
+	sub := claims["sub"].(string)
+	iss := claims["iss"].(string)
+
+	if os.Getenv("OIDC_ISSUER") == "" || iss != os.Getenv("OIDC_ISSUER") {
+		er := models.LicenseError{
+			Status:    http.StatusUnauthorized,
+			Message:   "Please check your credentials and try again",
+			Error:     "Please check your credentials and try again",
+			Path:      c.Request.URL.Path,
+			Timestamp: time.Now().Format(time.RFC3339),
+		}
+		c.JSON(http.StatusUnauthorized, er)
+		log.Printf("\033[31mError: Issuer '%s' not supported\033[0m", iss)
+		return
+	}
+
+	email := claims[os.Getenv("OIDC_EMAIL_KEY")].(string)
+	username := claims[os.Getenv("OIDC_USERNAME_KEY")].(string)
+	level := "USER"
+
+	var user, updatedUser models.User
+	user = models.User{
+		Username:  &username,
+		UserEmail: &email,
+		Userlevel: &level,
+		OidcSub:   &sub,
+		OidcIss:   &iss,
+	}
+
+	result := db.DB.
+		Where(&models.User{Username: user.Username}).
+		FirstOrCreate(&user)
+
+	if result.Error != nil {
+		errMessage := "Something went wrong. Try again."
+		if errors.Is(result.Error, gorm.ErrDuplicatedKey) {
+			if err := db.DB.Where(&models.User{UserEmail: &email}).First(&user).Error; err != nil {
+				er := models.LicenseError{
+					Status:    http.StatusInternalServerError,
+					Message:   "Something went wrong",
+					Error:     err.Error(),
+					Path:      c.Request.URL.Path,
+					Timestamp: time.Now().Format(time.RFC3339),
+				}
+				c.JSON(http.StatusInternalServerError, er)
+				return
+			}
+			if user.OidcSub == nil {
+				// Case when user is logged in with email password and wants to log in with OIDC with the same email
+				// TODO: Add changelogs for user
```

Comment:
You want to implement this todo here or in a new PR?

Reply:
New PR